test_plans/sub_group.asciidoc

```diff
@@ -0,0 +1,44 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for sub-group class
+
+This is a test plan for the sub-group class as described in Section 4.9.1.8. of the SYCL 2020 specification.
+
+Estimated development time is three days.
+
+== Testing scope
+
+=== Backend coverage
+
+All the tests described below are not backend-specific and are performed for any SYCL backend.
+
+=== Device coverage
+
+All tests construct a test device for which conformance is assessed. All the tests described below are performed once for that test device.
+
+== Tests
+A kernel is defined that calls all functions below, checks their return type and records the returned value in a list of a size equal to the number of work-items times the number of functions. The kernel is called with `parallel_for(nd_range<1>(range<1>(work_group_size), range<1>(work_group_size)), [](nd_item<1> item){})`, let `sub_group` be `item.get_sub_group()`:
```

Comment:
I suppose that would be possible but, reading the specification again, I am confused about how partitioning into sub-groups is performed. The spec states for `get_local_range()` (emphasis mine):
> Return a range representing the size of the sub-group. This size may be less than the value returned by `get_max_local_range()`, depending on the position of the sub-group within its parent work-group **and the manner in which sub-groups are constructed by the implementation**.

In the test plan, as it stands, I assume that the first $n$ work-items of a one-dimensional `nd_range` get placed in the first sub-group, the next $n$ in the second sub-group, etc. I doubt that there is a justification for an implementation to do it differently. But it is not clear to me how that would work for higher-dimensional `nd_ranges`, and if that is specified by the specification at all. This also relates to your question below https://github.com/KhronosGroup/SYCL-CTS/pull/436#discussion_r1054630603.

Reply:
> In the test plan, as it stands, I assume that the first $n$ work-items of a one-dimensional `nd_range` get placed in the first sub-group, the next $n$ in the second sub-group, etc.

I think this is not a valid assumption.  The spec specifically says that the number of work-items in each sub-group is implementation defined.  See this in [section 3.9.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels):

> The size and number of sub-groups is implementation-defined and may differ for each kernel, and different devices may make different guarantees with respect to how sub-groups within a work-group are scheduled. The maximum number of work-items in any sub-group in a kernel is based on a combination of the kernel and its dispatch dimensions. The size of any sub-group in the dispatch is between 1 and this maximum sub-group size, and the size of an individual sub-group is invariant for the duration of a kernel’s execution.

The more I think about it, the more I think that my proposed checks in [this comment](https://github.com/KhronosGroup/SYCL-CTS/pull/436#discussion_r1054621832) are the proper way to test the `sub_group` class.  We need to be careful that the CTS follows the letter of the spec and does not make assumptions about what a "reasonable implementation" should do.

This discussion has made me wonder whether the `[sycl::reqd_sub_group_size]` attribute should provide some stronger guarantee.  For example, maybe that attribute should require the work-group size to be evenly divisible by the required sub-group size (in each dimension), and then the spec could guarantee that every sub-group has exactly the requested size.  Tagging @Pennycook for comment on this.  Note, though, that the spec does *not* provide this guarantee now, so we should not test for this in the CTS.

Reply:
You raise valid points, in the other comment thread as well. I agree that fully relying on `sycl::reqd_sub_group_size` for the sub-group tests is flawed. As such, I have updated the test plan to include this test when the sub-group size is not known beforehand.

Do you think it is best to also drop the part of the test with `sycl::reqd_sub_group_size`? If I understand it correctly, the expected values for `subgroup.get_group_id`, `subgroup.get_local_id`, etc. cannot be calculated, even if the sub-group size is known beforehand due to the fact that the implementation may choose how sub-groups are constructed.

Reply:
To illustrate the latter point:
```
work-group range: (2,3)
sub-group size:   2

division might be:
[(0,0),(0,1)],[(0,2),(1,0)],[(1,1),(1,2)]
with local_ids: 0, 1, 0, 1, 0, 1

or, it might be:
[(0,0),(0,1)],[(0,2)],[(1,0),(1,1)],[(1,2)]
with local_ids: 0, 1, 0, 0, 1, 0
```
The purpose of the test that enforced the sub-group size, is that the expected values of the member functions of the sub-group class can be calculated. Since this seems not to be the case, in my opinion it is best to leave this test out of the plan.

Reply:
BTW, I captured some of the open questions that have arisen from this test plan here: https://gitlab.khronos.org/sycl/Specification/-/issues/638

Reply:
Okay, I have removed this part from the test plan.